### PR TITLE
Fragments Part 1

### PR DESCRIPTION
**Pull request by jws-1**
Initial fragments PR (see #1247).

This PR implements the following:
- Ability to define a `FragmentSite` using a NETA string, sites are then determined by matching atoms to this definition (while preserving uniqueness).
- Use of identifiers to assign atoms that contribute to the origin, x and y axis.
- I/O for `FragmentSite`s.
- Fixed a bug with the input file parser, where `#`s inside string literals were seen as comments.
- A (temporarily disabled, until further work is done) system test that refactors the RDF water example to utilise fragment sites.

A visual test can be performed by loading `tests/siterdf/rdf_fragments.txt`, and viewing the sites in the species viewer.

### CODE REVIEW

File: src/neta/neta.h [hunk 1]
```diff
@@ -45,6 +47,10 @@ class NETADefinition
     std::string_view definitionString() const;
     // Return whether the definition is valid
     bool isValid() const;
+    // Add identifiers
+    void addIdentifier(std::string identifier);
+    // Return identifiers from definition
+    std::set<std::string> identifiers();
```
Comment (trisyoungs) [inline src/neta/neta.h:53]: Should this be:
```suggestion
    const std::set<std::string> &identifiers() const;
```

**Review by trisyoungs — APPROVED** [review]
Really good. I have nothing significant to correct so have approved, but have suggested one significant optimisation to the whole process that I think is worth addressing. Could be done as part of this PR but, on reflection, might be best left to a follow-up PR since it touches beyond the new functionality.

File: src/classes/sitestack.cpp [hunk 2]
```diff
@@ -188,6 +188,168 @@ bool SiteStack::createDynamic()
     return true;
 }
 
+bool SiteStack::createFragmentOriented()
```
[the remark this answers is not in the document]
Comment (jws-1) [reply src/classes/sitestack.cpp:191]: That makes sense, considering the order of prototypes in the header file!

File: src/classes/sitestack.cpp [hunk 3]
```diff
@@ -188,6 +188,168 @@ bool SiteStack::createDynamic()
...
+            // Check if the fragment we have found is unique.
+            std::sort(matchedAtomIndices.begin(), matchedAtomIndices.end());
+            if (std::find(matchedIndices.begin(), matchedIndices.end(), matchedAtomIndices) != matchedIndices.end())
+                continue;
+
+            // If it's unique, remember it and proceed.
+            matchedIndices.push_back(std::move(matchedAtomIndices));
+            siteIndices.push_back(i.index());
+        }
+    }
```
[the remark this answers is not in the document]
Comment (jws-1) [reply src/classes/sitestack.cpp:319]: Yes, I agree. It definitely would improve performance if we can get away with only doing this stuff once!

File: src/classes/speciessite.cpp [hunk 4]
```diff
@@ -131,9 +137,9 @@ std::vector<int> SpeciesSite::originAtomIndices() const
 // Set whether the origin should be calculated with mass-weighted positions
 void SpeciesSite::setOriginMassWeighted(bool b)
 {
-    if (type_ != SpeciesSite::SiteType::Static)
+    if (type_ != SpeciesSite::SiteType::Static && type_ != SpeciesSite::SiteType::Fragment)
     {
-        Messenger::error("Setting mass weighting for a non-static site is not permitted.\n");
+        Messenger::error("Setting mass weighting for a non-static or non-fragment site is not permitted.\n");
```
[the remark this answers is not in the document]
Comment (jws-1) [reply src/classes/speciessite.cpp:142]: I assumed (with some reasoning that I cannot recall) that we could determine the origin by mass weighting for fragments sites?

File: src/neta/neta.h (hunk 1 above)
Comment (jws-1) [reply src/neta/neta.h:53]: Yes, definitely.

**Comment by jws-1** [comment]
> 

I agree with the suggested optimisations - I will capture this in an issue to be addressed separately.

**Comment by jws-1** [comment]
> Really good. I have nothing significant to correct so have approved, but have suggested one significant optimisation to the whole process that I think is worth addressing. Could be done as part of this PR but, on reflection, might be best left to a follow-up PR since it touches beyond the new functionality.

I opened an issue, #1463, feel free to add anything you think is missing!

File: src/classes/speciessite.cpp (hunk 4 above)
Comment (trisyoungs) [inline src/classes/speciessite.cpp:142]: Sorry! Got my logic in a twist here! There's not really a reason why you couldn't select this option for the `Dynamic` sites - it just doesn't do anything.  We can probably just remove this check.

[the remark this answers is not in the document]
Comment (trisyoungs) [reply src/classes/speciessite.cpp:142]: Sorry! Got my logic in a twist here! There's not really a reason why you couldn't select this option for the `Dynamic` sites - it just doesn't do anything.  We can probably just remove this check.

[the remark this answers is not in the document]
Comment (jws-1) [reply src/classes/speciessite.cpp]: OK - I've dropped that check!

**Pull request merged**